CLUSTER.md

```diff
@@ -0,0 +1,450 @@
+of the total nodes in the cluster, receives & propagates heartbeat information from them and enables them to make failover decisions in a
+consistent & timely manner.
+
+Maintaining these two tiers allows scaling the number of nodes while:
+- Maintaining an upper bound over the time required to decide and propagate failover decisions.
+While the time required to propagate the updated primary/replica status depends on the number of Failover Coordinators in the deployment
+the maximum number of these should always be small enough to allow us to guarantee a substantially better SLA then what
+the current implementation provides.
+- Limiting the clustering overhead on individual nodes in the deployment.
+Depending on pub/sub mode, an individual node only needs to maintain a connection to its Failover Coordinator (and primary/replicas).
+
+While Flotilla can be implemented using any sort of strongly consistent system, the reference implementation we propose
+uses Redis to host the Topology Director & Failover Coordinator logic. This allows for different deployment modes depending on 
+the needs and orchestrating abilities of the administrator:
+- Full Blown: Dedicated nodes holding the Topology Directors and multiple Failover Coordinators dividing the data nodes among them.
+- Midrange: Dedicated nodes holding both the Topology Director and Failover Coordinator (with just one Failover Coordinator owning all the data nodes).
+- Lightweight: Topology Director and Failover Coordinator located on a subset of data nodes. This removes the need to maintain and orchestrate
+non-data nodes at the cost of potentially impacting performance on the dual-use nodes.
+- Naive: Topology Director and Failover Coordinator located on all nodes. This allows for the quickest bootstrapping, but severely limits scalability.
+
+![Flotilla High Level overview](images/high_level.jpg?raw=true "High Level Overview")
+
+### The Topology Director
+
+The Topology Director serves as the source of truth for the cluster topology (excluding primary/replica status) and helps propagate
+information from Failover Coordinators. It never handles any user data, only topology metadata.
+Since it owns the metadata of the entire cluster it can allow changing this metadata while preserving the strong
+consistency guarantees we desire.
+For example - it can ensure that if a Shard holds a slot marked as Migrating, there's a corresponding Shard with the slot marked as Importing.
+We also rely on the Failover Coordinator to distribute the primary/replica state of individual nodes - This state is received periodically
+from every Failover Coordinator, cached in the Topology Director and relayed to all other Failover Coordinators when they send their state.
+
+- A completely reactive entity
+- Receives requests to alter the cluster topology and adjusts the internal state as needed
+- Validates change requests for correctness
+- Responds to topology queries by the other components
+
+#### Interfaces
```

Comment:
We can split the API into three different segments:
- Pure data path (which isn't going to change and isn't outlined here)
- Control path which needs to be supported by any data node (Join, Heartbeat, maybe topology)
- Control path which interacts with the FC/TD, which needs to be part of the spec since we aim to publish a complete solution, but could be replaced by an alternative cluster implementation without any impact on data nodes.

Reply:
> Control path which interacts with the FC/TD, which needs to be part of the spec since we aim to publish a complete solution, but could be replaced by an alternative cluster implementation without any impact on data nodes.

Sure, we plan on publishing a complete solution, but it's not really part of the spec.

Reply:
I also just realized I might be using "spec" differently than you. I've been thinking about is as a "generic" set of APIs other people could build against, where you just might be using it to describe what we plan to build. If that is the case, I can rephrase some of these comments since they probably don't make sense :)

Reply:
To make sure we're aligned - the APIs here specify all three segments (Control, Control+Data, Data) of Flotilla.
As part of working on the reference implementation we will need to make sure tools like redis-cli know how to work with all of these.

Like we discussed, it should be possible to create an alternative implementation which completely changes some or all of the "Control" APIs without touching the other two segments. Non admin Redis clients (clients which do not modify the control layer) are supposed to be completely fine with such an alternative implementation.

Reply:
Resolving this as it is essentially discussing the same topic as the thread below (https://github.com/redis/redis/pull/10875#discussion_r900329927)